clients/xset/xset.man
.TH XSET 1 "25 October 1988" "X Version 11"
.SH NAME
xset - user preference utility for X
.SH SYNOPSIS
.B xset
[-display \fIdisplay\fP]
[-b] [b on/off] [b [\fIvolume\fP [\fIpitch\fP [\fIduration\fP]]]
[-c] [c on/off] [c [\fIvolume\fP]]
[[-+]fp[-+=] \fIpath\fP[,\fIpath\fP[,...]]] [fp default] [fp rehash]
[[-]led [\fIinteger\fP]] [led on/off]
[m[ouse] [\fIacceleration\fP [\fIthreshold\fP]]] [m[ouse] default]
[p \fIpixel\fP \fIcolor\fP]
[[-]r] [r on/off]
[s [\fIlength\fP [\fIperiod\fP]]] [s blank/noblank]
[s expose/noexpose] [s on/off] [s default]
[q]
.SH DESCRIPTION
This program is used to set various user preference options of the display.
.SH OPTIONS
.PP
.TP 8
.B \-display \fIdisplay\fP
This option specifies the server to use; see \fIX(1)\fP.
.PP
.TP 8
.B b
the \fBb\fP option controls bell volume, pitch and duration.
This option accepts up to three numerical parameters, a preceding
dash(-), or a 'on/off' flag.  If no parameters are
given, or the 'on' flag is used, the system defaults will be used.
If the dash or 'off' are given, the bell will be turned
off.
If only one numerical parameter is given, the
bell volume will be set to that value, as a percentage of its maximum.
Likewise, the second numerical
parameter specifies the bell pitch, in hertz, and
the third numerical parameter
specifies the duration in milliseconds.  Note that not
all hardware can vary the bell characteristics.  The X server will set
the characteristics of the bell as closely as it can to the user's
specifications.
.PP
.TP 8
.B c
The \fBc\fP option controls key click.
This option can take an optional value, a preceding dash(-),
or an 'on/off' flag.
If no parameter or the 'on' flag is given, the system defaults
will be used. If the dash or 'off' flag is used, keyclick will be
disabled.
If a value from 0 to 100 is given, it is used to
indicate volume, as a percentage of the maximum.
The X server will set
the volume to the nearest value that the hardware can support.
.PP
.TP 8
.B fp= \fIpath,...\fP
The \fBfp=\fP sets the font path to the directories given in the path argument.
The directories are interpreted by the server, not by the client, and are
server-dependent.  Directories that do not contain font databases created by
\fImkfontdir\fP will be ignored by the server.
.TP 8
.B fp \fIdefault\fP
The \fIdefault\fP argument causes the font path to be reset to the server's
default.
.TP 8
.B fp \fIrehash\fP
The \fIrehash\fP argument causes the server to reread the font databases in
the current font path.  This is generally only used when adding new fonts to
a font directory (after running \fImkfontdir\fP to recreate the font database).
.PP
.TP 8
.B "\-fp \fRor\fP fp\-"
The \fB\-fp\fP and \fBfp\-\fP options remove elements from the current
font path.  They must be followed by a comma-separated list of directories.
.PP
.TP 8
.B "\+fp \fRor\fP fp\+"
This \fB\+fp\fP and \fBfp\+\fP options prepend and append elements to the 
current font path, respectively.  They must be followed by a comma-separated 
list of directories.
.PP
.TP 8
.B led
The \fBled\fP option controls the keyboard LEDs.
This controls the turning on or off of one or all of the LEDs.
It accepts an optional integer, a preceding dash(-) or an 'on/off' flag.
If no parameter or the 'on' flag is given, all LEDs are turned on.
If a preceding dash or the flag 'off' is given, all LEDs are turned off.
If a value between 1 and 32 is given, that LED will be turned on or off
depending on the existance of a preceding dash.
A common LED which can be controlled is the ``Caps Lock'' LED.  ``xset
led 3'' would turn led #3 on.  ``xset -led 3'' would turn it off.
The particular LED values may refer to different LEDs on different
hardware.
.PP
.TP 8
.B m
The \fBm\fP option controls the mouse parameters.
The parameters for the mouse are `acceleration' and `threshold'.
The mouse, or whatever pointer the machine is connected to,
will go `acceleration' times as fast when it travels more than `threshold'
pixels in a short time.  This way, the mouse can be used for precise
alignment when it is moved slowly, yet it can be set to travel across
the screen in a flick of the wrist when desired.  One or both
parameters for the 
.B m
option can be omitted, but if only one is
given, it will be interpreted as the acceleration.
If no parameters or the flag 'default' is used, the system defaults will
be set.
.PP
.TP 8
.B p
The \fBp\fP option controls pixel color values.
The parameters are the color map entry number in decimal,
and a color specification.  The root background colors may be changed
on some servers by altering the entries for BlackPixel and WhitePixel.
Although these are often 0 and 1, they need not be.  Also, a server may
choose to allocate those colors privately, in which case an error will
be generated.  The map entry must not be a read-only color,
or an error will result.
.PP
.TP 8
.B r
The \fBr\fP option controls the autorepeat.
If a preceding dash or the 'off' flag is used, autorepeat will be disabled.
If no parameters or the 'on' flag is used, autorepeat will be enabled.
.PP
.TP 8
.B s
The \fBs\fP option lets you set the screen saver parameters.
This option accepts up to two numerical parameters, a 'blank/noblank'
flag, an 'expose/noexpose' flag, an 'on/off' flag, or the 'default' flag.
If no parameters or the 'default' flag is used, the system will be set
to its default screen saver characteristics.
The 'on/off' flags simply turn the screen saver functions on or off.
The 'blank' flag sets the
preference to blank the video (if the hardware can do so) rather than
display a background pattern, while 'noblank' sets the
preference to display a pattern rather than blank the video.
The 'expose' flag sets the
preference to allow window exposures (the server can freely discard
window contents), while 'noexpose' sets the preference to disable
screen saver unless the server can regenerate the screens without
causing exposure events.
The length and period
parameters for the screen saver function determines how long the
server must be inactive for screen saving to activate, and the period
to change the background pattern to avoid burn in.
The arguments are specified in seconds.
If only one numerical parameter is given, it will be used for the length.
.PP
.TP 8
.B q
The \fBq\fP option gives you information on the current settings.
.PP
These settings will be reset to default values when you log out.
.PP
Note that not all X implementations are guaranteed to honor all of these
options.
.SH "SEE ALSO"
X(1), Xserver(1), xmodmap(1), xrdb(1), xsetroot(1)
.SH COPYRIGHT
Copyright 1988, Massachusetts Institute of Technology.
.br
See \fIX(1)\fP for a full statement of rights and permissions.
.SH AUTHOR
Bob Scheifler, MIT Laboratory for Computer Science
.br
David Krikorian, MIT Project Athena (X11 version)


CHANGES
The following changes have been made to this directory since R2:

     o  substantially improved the X man page.
     o  added README file with quick instructions for those in the fast lane.
     o  added RELEASENOTES file giving overview of what has changed and where
	to look for more information.
     o  added STANDARDS file describing what is and is not part of the X
	Standard.
     o  removed support for automatically building contrib directory from
	the Imakefile since it won't work completely on any single platform.
     o  moved rgb directory here from util.
     o  added some misc. utility targets to the top-level Makefile.


clients/xset/xset.c
/* 
 * $XConsortium: xset.c,v 1.46 88/10/15 18:35:26 jim Exp $ 
 */
#include <X11/copyright.h>

/* Copyright    Massachusetts Institute of Technology    1985	*/

#ifndef lint
static char *rcsid_xset_c = "$XConsortium: xset.c,v 1.46 88/10/15 18:35:26 jim Exp $";
#endif

#include <X11/Xos.h>
#include <X11/Xlib.h>
/*  #include <X11/Xlibwm.h>  [Doesn't exist yet  5-14-87]  %*/
#include <X11/keysym.h>
#include <X11/Xproto.h>
#include <X11/Xutil.h>
#include <X11/Xmu.h>
#include <stdio.h>
#include <ctype.h>

#define ON 1
#define OFF 0

#define SERVER_DEFAULT (-1)
#define DONT_CHANGE -2

#define ALL -1
#define TIMEOUT 1
#define INTERVAL 2
#define PREFER_BLANK 3
#define ALLOW_EXP 4

#define	nextarg(i, argv) \
	argv[i]; \
	if (i >= argc) \
		break; \

char *progName;

int local_xerror();

main(argc, argv)
int argc;
char **argv;
{
register char *arg;
register int i;
int percent;
int acc_num, acc_denom, threshold;
XKeyboardControl values;
unsigned long pixels[512];
caddr_t colors[512];
XColor def;
int numpixels = 0;
char *disp = NULL;
Display *dpy;
progName = argv[0];
if (argc == 1)  usage (NULL, NULL);    /* To be replaced by window-interface */
for (i = 1; i < argc; i++) {
  arg = argv[i];
  if (strcmp (arg, "-display") == 0 || strcmp (arg, "-d") == 0) {
    if (++i >= argc) usage ("missing argument to -display", NULL);
    disp = argv[i];
  } 
}
dpy = XOpenDisplay(disp);  /*  Open display and check for success */
if (dpy == NULL) {
  fprintf(stderr, "%s:  unable to open display \"%s\"\n",
	  argv[0], XDisplayName (disp));
  exit(1);
}
XSetErrorHandler (local_xerror);
for (i = 1; i < argc; ) {
  arg = argv[i++];
  if (strcmp (arg, "-display") == 0 || strcmp (arg, "-d") == 0) {
    ++i;					/* already dealt with */
    continue;
  } else if (*arg == '-' && *(arg + 1) == 'c'){ /* Does arg start with "-c"? */
    set_click(dpy, 0);           /* If so, turn click off and  */
  } 
  else if (*arg == 'c') {         /* Well, does it start with "c", then? */
    percent = SERVER_DEFAULT;		/* Default click volume. */
    if (i >= argc) {
	set_click (dpy, percent);	/* set click to default */
	break;
    }
    arg = nextarg(i, argv);
    if (strcmp(arg, "on") == 0) {               /* Let click be default. */
      i++;
    } 
    else if (strcmp(arg, "off") == 0) {  
      percent = 0;       /* Turn it off.          */
      i++;
    } 
    else if (isnumber(arg, 100)) {
      percent = atoi(arg);  /* Set to spec. volume */
      i++;
    }
    set_click(dpy, percent);
  } 
  else if (*arg == '-' && *(arg + 1) == 'b') {  /* Does arg start w/ "-b" */
    set_bell_vol(dpy, 0);           /* Then turn off bell.    */
  } 
  else if (*arg == 'b') {                       /* Does it start w/ "b".  */
    percent = SERVER_DEFAULT;		/* Set bell to default. */
    if (i >= argc) {
	set_bell_vol (dpy, percent);	/* set bell to default */
	break;
    }
    arg = nextarg(i, argv);
    if (strcmp(arg, "on") == 0) {               /* Let it stay that way.  */
      set_bell_vol(dpy, percent);
      i++;
    } 
    else if (strcmp(arg, "off") == 0) {
      percent = 0;            /* Turn the bell off.     */
      set_bell_vol(dpy, percent);
      i++;
    } 
    else if (isnumber(arg, 100)) {              /* If volume is given:    */
      percent = atoi(arg);    /* set bell appropriately.*/
      set_bell_vol(dpy, percent);
      i++;
      arg = nextarg(i, argv);

      if (isnumber(arg, 20000)) {               /* If pitch is given:     */
	set_bell_pitch(dpy, atoi(arg));    /* set the bell.           */
	i++;

	arg = nextarg(i, argv);
	if (isnumber(arg, 1000)) {              /* If duration is given:  */
	  set_bell_dur(dpy, atoi(arg));  /*  set the bell.      */
	  i++;
	}
      }
    }
    else
      set_bell_vol (dpy, percent);		/* set bell to default */
  }
  else if (strcmp(arg, "fp") == 0) {	       /* set font path */
    if (i >= argc) {
	arg = "default";
    } else {
	arg = nextarg(i, argv);
    }
    set_font_path(dpy, arg, 1, 0, 0);	/* special argument */
    i++;
  }
  else if (strcmp(arg, "fp=") == 0) {	/* unconditionally set */
    if (i >= argc) {
	usage ("missing fp= argument", NULL);
    } else {
	arg = nextarg(i, argv);
    }
    set_font_path(dpy, arg, 0, 0, 0);	/* not special, set */
    i++;
  }
  else if (strcmp(arg, "+fp") == 0) {	       /* set font path */
    if (i >= argc) usage ("missing +fp argument", NULL);
    arg = nextarg(i, argv);
    set_font_path(dpy, arg, 0, 1, 0);	/* not special, prepend */
    i++;
  }
  else if (strcmp(arg, "fp+") == 0) {	       /* set font path */
    if (i >= argc) usage ("missing fp+ argument", NULL);
    arg = nextarg(i, argv);
    set_font_path(dpy, arg, 0, 0, 1);	/* not special, append */
    i++;
  }
  else if (strcmp(arg, "-fp") == 0) {	       /* set font path */
    if (i >= argc) usage ("missing -fp argument", NULL);
    arg = nextarg(i, argv);
    set_font_path(dpy, arg, 0, -1, 0);	/* not special, preremove */
    i++;
  }
  else if (strcmp(arg, "fp-") == 0) {	       /* set font path */
    if (i >= argc) usage ("missing fp- argument", NULL);
    arg = nextarg(i, argv);
    set_font_path(dpy, arg, 0, 0, -1);	/* not special, postremove */
    i++;
  }
  else if (strcmp(arg, "-led") == 0) {         /* Turn off one or all LEDs */
    values.led_mode = OFF;
    values.led = ALL;        /* None specified */
    arg = nextarg(i, argv);
    if (isnumber(arg, 32) && atoi(arg) > 0) {
      values.led = atoi(arg);
      i++;
    }
    set_led(dpy, values.led, values.led_mode);
  } 
  else if (strcmp(arg, "led") == 0) {         /* Turn on one or all LEDs  */
    values.led_mode = ON;
    values.led = ALL;
    if (i >= argc) {
	set_led (dpy, values.led, values.led_mode);	/* set led to def */
	break;
    }
    arg = nextarg(i, argv);
    if (strcmp(arg, "on") == 0) {
      i++;
    } 
    else if (strcmp(arg, "off") == 0) {       /*  ...except in this case. */
       values.led_mode = OFF;
      i++;
    }
    else if (isnumber(arg, 32) && atoi(arg) > 0) {
      values.led = atoi(arg);
      i++;
    }
    set_led(dpy, values.led, values.led_mode);
  }
/*  Set pointer (mouse) settings:  Acceleration and Threshold. */
  else if (strcmp(arg, "m") == 0 || strcmp(arg, "mouse") == 0) {
    acc_num = SERVER_DEFAULT;		/* restore server defaults */
    acc_denom = SERVER_DEFAULT;
    threshold = SERVER_DEFAULT;
    if (i >= argc){
      set_mouse(dpy, acc_num, acc_denom, threshold);
      break;
    }
    arg = argv[i];
    if (strcmp(arg, "default") == 0) {
      i++;
    } 
    else if (*arg >= '0' && *arg <= '9') {
      acc_num = atoi(arg);  /* Set acceleration to user's tastes.  */
      i++;
      if (i >= argc) {
	set_mouse(dpy, acc_num, acc_denom, threshold);
	break;
      }
      arg = argv[i];
      if (*arg >= '0' && *arg <= '9') {
	threshold = atoi(arg);  /* Set threshold as user specified.  */
	i++;
      }
    }
      set_mouse(dpy, acc_num, acc_denom, threshold);
  } 
  else if (*arg == 's') {   /*  If arg starts with "s".  */
    if (i >= argc) {
      set_saver(dpy, ALL, 0);  /* Set everything to default  */
      break;
    }
    arg = argv[i];
    if (strcmp(arg, "blank") == 0) {       /* Alter blanking preference. */
      set_saver(dpy, PREFER_BLANK, PreferBlanking);
      i++;
    }
    else if (strcmp(arg, "noblank") == 0) {     /*  Ditto.  */
      set_saver(dpy, PREFER_BLANK, DontPreferBlanking);
      i++;
    }
    else if (strcmp(arg, "expose") == 0) {       /* Alter exposure preference. */
      set_saver(dpy, ALLOW_EXP, AllowExposures);
      i++;
    }
    else if (strcmp(arg, "noexpose") == 0) {     /*  Ditto.  */
      set_saver(dpy, ALLOW_EXP, DontAllowExposures);
      i++;
    }
    else if (strcmp(arg, "off") == 0) {
      set_saver(dpy, TIMEOUT, 0);   /*  Turn off screen saver.  */
      i++;
      if (i >= argc)
	break;
      arg = argv[i];
      if (strcmp(arg, "off") == 0) {
	set_saver(dpy, INTERVAL, 0);
	i++;
      }
    }
    else if (strcmp(arg, "default") == 0) {    /*  Leave as default.       */
      set_saver(dpy, ALL, 0);
      i++;
    } 
    else if (*arg >= '0' && *arg <= '9') {  /*  Set as user wishes.   */
      set_saver(dpy, TIMEOUT, atoi(arg));
      i++;
      if (i >= argc)
	break;
      arg = argv[i];
      if (*arg >= '0' && *arg <= '9') {
	set_saver(dpy, INTERVAL, atoi(arg));
	i++;
      }
    }
  } 
  else if(*arg == '-' && *(arg + 1) == 'r'){ /* If arg starts w/ "-r" */
    set_repeat(dpy, ALL, OFF);
  } 
  else if (*arg == 'r') {            /*  If it starts with "r"        */
    if (i >= argc) {
      set_repeat(dpy, ALL, ON);
      break;
    }
    arg = argv[i];                   /*  Check next argument.         */
    if (strcmp(arg, "on") == 0) {
      set_repeat(dpy, ALL, ON);
      i++;
    } 
    else if (strcmp(arg, "off") == 0) {
      set_repeat(dpy, ALL, OFF);
      i++;
    }
  } 
  else if (*arg == 'p') {           /*  If arg starts with "p"       */
    if (i + 1 >= argc)
      usage ("missing argument to p", NULL);
    arg = argv[i];
    if (*arg >= '0' && *arg <= '9')
      pixels[numpixels] = atoi(arg);
    else
      usage ("invalid pixel number %s", arg);
    i++;
    colors[numpixels] = argv[i];
    i++;
    numpixels++;
  }
  else if (*arg == '-' && *(arg + 1) == 'k') {
    set_lock(dpy, OFF);
  }
  else if (*arg == 'k') {         /*  Set modifier keys.               */
    set_lock(dpy, ON);
  }
  else if (*arg == 'q' || strncmp(arg, "-q", 2) == 0) {	 /* query status */
    query(dpy);
  }
  else
    usage ("unknown option %s", arg);
}

if (numpixels)
    set_pixels(dpy, pixels, colors, numpixels);

XCloseDisplay (dpy);

exit(0);    /*  Done.  We can go home now.  */
}


isnumber(arg, maximum)
	char *arg;
	int maximum;
{
	register char *p;

	if (arg[0] == '-' && arg[1] == '1' && arg[2] == '\0')
		return(1);
	for (p=arg; isdigit(*p); p++);
	if (*p || atoi(arg) > maximum)
		return(0); 
	return(1);
}

/*  These next few functions do the real work (xsetting things).
 */
set_click(dpy, percent)
Display *dpy;
int percent;
{
XKeyboardControl values;
values.key_click_percent = percent;
XChangeKeyboardControl(dpy, KBKeyClickPercent, &values);
return;
}

set_bell_vol(dpy, percent)
Display *dpy;
int percent;
{
XKeyboardControl values;
values.bell_percent = percent;
XChangeKeyboardControl(dpy, KBBellPercent, &values);
return;
}

set_bell_pitch(dpy, pitch)
Display *dpy;
int pitch;
{
XKeyboardControl values;
values.bell_pitch = pitch;
XChangeKeyboardControl(dpy, KBBellPitch, &values);
return;
}

set_bell_dur(dpy, duration)
Display *dpy;
int duration;
{
XKeyboardControl values;
values.bell_duration = duration;
XChangeKeyboardControl(dpy, KBBellDuration, &values);
return;
}

/*
 * Set, add, or subtract the path according to before and after flags:
 *
 *	before	after	action
 *
 *	   0      0	FontPath := path
 *	  -1      0	FontPath := current - path
 *	   0     -1	FontPath := current - path
 *	   1      0	FontPath := path + current
 *	   0      1	FontPath := current + path
 */

set_font_path(dpy, path, special, before, after)
    Display *dpy;
    char *path;
    int special, before, after;
{
    char **directoryList = NULL; int ndirs = 0;
    char **currentList = NULL; int ncurrent = 0;
    char **newList = NULL; int nnew = 0;

    if (special) {
	if (strcmp (path, "default") == 0) {
	    XSetFontPath (dpy, NULL, 0);
	    return;
	}
	if (strcmp (path, "rehash") == 0) {
	    currentList = XGetFontPath (dpy, &ncurrent);
	    if (!currentList) {
		fprintf (stderr, "%s:  unable to get current font path.\n",
			 progName);
		return;
	    }
	    XSetFontPath (dpy, currentList, ncurrent);
	    free ((char *) currentList);
	    return;
	} 
	/*
	 * for now, fall though and process keyword and directory list for
	 * compatibility with previous versions.
	 */
    }

    /*
     * parse the path list.  If before or after is non-zero, we'll need 
     * the current value.
     */

    if (before != 0 || after != 0) {
	currentList = XGetFontPath (dpy, &ncurrent);
	if (!currentList) {
	    fprintf (stderr, "%s:  unable to get old font path.\n",
		     progName);
	    before = after = 0;
	}
    }

    {
	/* count the number of directories in path */
	register char *cp = path;

	ndirs = 1;
	while ((cp = index (cp, ',')) != NULL) {
	    ndirs++;
	    cp++;
	}
    }

    directoryList = (char **) malloc (ndirs*sizeof (char *));
    if (!directoryList) error ("out of memory for font path directory list");

    {
	/* mung the path and set directoryList pointers */
	int i = 0;
	char *cp = path;

	directoryList[i++] = cp;
	while ((cp = index (cp, ',')) != NULL) {
	    directoryList[i++] = cp + 1;
	    *cp++ = '\0';
	}
	if (i != ndirs) {
	    fprintf (stderr, 
		     "%s: internal error, only parsed %d of %d directories.\n",
		     progName, i, ndirs);
	    exit (1);
	}
    }
	    
    /*
     * now we have have parsed the input path, so we can set it
     */

    if (before == 0 && after == 0) {
	XSetFontPath (dpy, directoryList, ndirs);
    }

    /* if adding to list, build a superset */
    if (before > 0 || after > 0) {
	int nnew = ndirs + ncurrent;
	char **newList = (char **) malloc (nnew * sizeof (char *));
	
	if (!newList) error ("out of memory");
	if (before > 0) {		/* new + current */
	    bcopy ((char *) directoryList, (char *) newList, 
		   (unsigned) (ndirs*sizeof (char *)));
	    bcopy ((char *) currentList, (char *) (newList + ndirs),
			(unsigned) (ncurrent*sizeof (char *)));
	    XSetFontPath (dpy, newList, nnew);
	} else if (after > 0) {
	    bcopy ((char *) currentList, (char *) newList,
		   (unsigned) (ncurrent*sizeof (char *)));
	    bcopy ((char *) directoryList, (char *) (newList + ncurrent),
		   (unsigned) (ndirs*sizeof (char *)));
	    XSetFontPath (dpy, newList, nnew);
	} 
	free ((char *) newList);
    }

    /* if deleting from list, build one the same size */
    if (before < 0 || after < 0) {
	int i, j;
	int nnew = 0;
	char **newList = (char **) malloc (ncurrent * sizeof (char *));
	
	if (!newList) error ("out of memory");
	for (i = 0; i < ncurrent; i++) {
	    for (j = 0; j < ndirs; j++) {
		if (strcmp (currentList[i], directoryList[j]) == 0)
		  break;
	    }
	    /* if we ran out, then insert into new list */
	    if (j == ndirs) newList[nnew++] = currentList[i];
	}
	if (nnew == ncurrent) {
	    fprintf (stderr,
		     "%s:  warning, no entries deleted from font path.\n",
		     progName);
	}
	XSetFontPath (dpy, newList, nnew);
	free ((char *) newList);
    }

    if (directoryList) free ((char *) directoryList);
    if (currentList) free ((char *) currentList);

    return;
}


set_led(dpy, led, led_mode)
Display *dpy;
int led, led_mode;
{
  XKeyboardControl values;
  values.led_mode = led_mode;
  if (led != ALL) {
    values.led = led;
    XChangeKeyboardControl(dpy, KBLed | KBLedMode, &values);
  }
  else {
    XChangeKeyboardControl(dpy, KBLedMode, &values);
  }
  return;
}

set_mouse(dpy, acc_num, acc_denom, threshold)
Display *dpy;
int acc_num, acc_denom, threshold;
{
int do_accel = True, do_threshold = True;

if (acc_num == DONT_CHANGE)		/* what an incredible crock... */
  do_accel = False;
if (threshold == DONT_CHANGE)
  do_threshold = False;
if (acc_num < 0) 			/* shouldn't happen */
  acc_num = SERVER_DEFAULT;
if (acc_denom <= 0)			/* prevent divide by zero */
  acc_denom = SERVER_DEFAULT;
if (threshold < 0) threshold = SERVER_DEFAULT;
XChangePointerControl(dpy, do_accel, do_threshold, acc_num,
		      acc_denom, threshold);
return;
}

set_saver(dpy, mask, value)
Display *dpy;
int mask, value;
{
  int timeout, interval, prefer_blank, allow_exp;
  XGetScreenSaver(dpy, &timeout, &interval, &prefer_blank, 
		  &allow_exp);
  if (mask == TIMEOUT) timeout = value;
  if (mask == INTERVAL) interval = value;
  if (mask == PREFER_BLANK) prefer_blank = value;
  if (mask == ALLOW_EXP) allow_exp = value;
  if (mask == ALL) {  /* "value" is ignored in this case.  (defaults) */
    timeout = SERVER_DEFAULT;
    interval = SERVER_DEFAULT;
    prefer_blank = DefaultBlanking;
    allow_exp = DefaultExposures;
  }
      XSetScreenSaver(dpy, timeout, interval, prefer_blank, 
		      allow_exp);
  return;
}

set_repeat(dpy, key, auto_repeat_mode)
Display *dpy;
int key, auto_repeat_mode;
{
  XKeyboardControl values;
  values.auto_repeat_mode = auto_repeat_mode;
  if (key != ALL) {
    values.key = key;
    XChangeKeyboardControl(dpy, KBKey | KBAutoRepeatMode, &values);
  }
  else {
    XChangeKeyboardControl(dpy, KBAutoRepeatMode, &values);
  }
  return;
}

set_pixels(dpy, pixels, colors, numpixels)
Display *dpy;
unsigned long pixels[512];
caddr_t colors[512];
int numpixels;
{
  XColor def;
  int scr = DefaultScreen (dpy);
  Visual *visual = DefaultVisual (dpy, scr);
  Colormap cmap = DefaultColormap (dpy, scr);
  int max_cells = DisplayCells(dpy, scr);
  XVisualInfo viproto, *vip;
  int nvisuals = 0;
  char *visual_type = NULL;
  int i;

  viproto.visualid = XVisualIDFromVisual (visual);
  vip = XGetVisualInfo (dpy, VisualIDMask, &viproto, &nvisuals);
  if (!vip || nvisuals != 1) {
      fprintf (stderr, "%s:  got %d visual info structures for visual 0x%lx\n",
	       progName, nvisuals, viproto.visualid);
      return;
  }

  switch (vip->class) {
      case GrayScale:
      case PseudoColor:
      case DirectColor:
	break;
      case StaticGray:
	visual_type = "StaticGray";
	break;
      case StaticColor:
	visual_type = "StaticColor";
	break;
      case TrueColor:
	visual_type = "TrueColor";
	break;
      default:
	fprintf (stderr, "%s:  unknown visual class type %d\n",
		 progName, vip->class);
	numpixels = 0;
  }

  if (visual_type) {
    fprintf (stderr, "%s:  cannot set pixel values in read-only %s visuals\n",
	     progName, visual_type);
  } else {
    for (i = 0; i < numpixels; i++) {
      def.pixel = pixels[i];
      if (def.pixel >= max_cells)
	fprintf(stderr, 
		"%s:  pixel value %d out of colormap range 0 through %d\n",
		progName, def.pixel, max_cells - 1);
      else {
        if (XParseColor (dpy, cmap, colors[i], &def))
	  XStoreColor(dpy, cmap, &def);
        else
	  fprintf (stderr, "%s:  invalid color \"%s\"\n", progName, colors[i]);
      }
    }
  }

  XFree ((char *) vip);

  return;
}

set_lock(dpy, onoff)
Display *dpy;
Bool onoff;
{
  XModifierKeymap *mods;
  mods = XGetModifierMapping(dpy);

  if (onoff)
    mods = XInsertModifiermapEntry(mods, XK_Caps_Lock, LockMapIndex);
  else
    mods = XDeleteModifiermapEntry(mods, XK_Caps_Lock, LockMapIndex);
  XSetModifierMapping(dpy, mods);
  return;
}

/*  This is the information-getting function for telling the user what the
 *  current "xsettings" are.
 */
query(dpy)
Display *dpy;
{
int scr = DefaultScreen (dpy);
XKeyboardState values;
int acc_num, acc_denom, threshold;
int timeout, interval, prefer_blank, allow_exp;
char **font_path; int npaths;

XGetKeyboardControl(dpy, &values);
XGetPointerControl(dpy, &acc_num, &acc_denom, &threshold);
XGetScreenSaver(dpy, &timeout, &interval, &prefer_blank, &allow_exp);
font_path = XGetFontPath(dpy, &npaths);

printf ("Keyboard Control Values:\n");
/*printf ("Auto Repeat:  %d \t\t", values.auto_repeat_mode);    %%*/
/*printf ("Key:  %d \n\n", values.key);     %%*/
printf ("Key Click Volume (%%):  %d \n", values.key_click_percent);
printf ("Bell Volume (%%):  %d \t", values.bell_percent);
printf ("Bell Pitch (Hz):  %d \t", values.bell_pitch);
printf ("Bell Duration (msec):  %d \n", values.bell_duration);
/*printf ("LED:  %d \t\t\t", values.led);
printf ("LED Mode:  %o \t\t", values.led_mode);         %%*/

printf ("Pointer Control Values:\n");
printf ("Acceleration:  %d (%d/%d)\t",
	acc_num / acc_denom, acc_num, acc_denom);
printf ("Threshold:  %d \n", threshold);
printf ("Screen Saver:\n");
printf ("Prefer Blanking:  %s\t",
	(prefer_blank == PreferBlanking) ? "Yes" :
	(prefer_blank == DontPreferBlanking) ? "No" :
	"<server violates spec>");
printf ("Allow Exposures:  %s\n",
	(allow_exp == AllowExposures) ? "Yes" :
	(allow_exp == DontAllowExposures) ? "No" :
	"<server violates spec>");
printf ("Time-out:  %d \t Cycle:  %d\n", timeout, interval);
printf ("Default colormap:   0x%lx \t BlackPixel:  %d \t WhitePixel:  %d\n",
	DefaultColormap (dpy, scr), 
	BlackPixel (dpy, scr), WhitePixel (dpy, scr));

if (npaths) {
    printf( "Font Path:  %s", *font_path++ );
    for( --npaths; npaths; npaths-- )
        printf( ",%s", *font_path++ );
    printf( "\n" );
} else {
    printf ("Font Path:  (empty)\n");
}
return;
}


/*  This is the usage function */

usage (fmt, arg)
    char *fmt;
    char *arg;
{
    if (fmt) {
	fprintf (stderr, "%s:  ", progName);
	fprintf (stderr, fmt, arg);
	fprintf (stderr, "\n\n");
    }

    fprintf (stderr, "usage:  %s [-display host:dpy] option ...\n", progName);
    fprintf (stderr, "    To turn bell off:\n");
    fprintf (stderr, "\t-b                b off               b 0\n");
    fprintf (stderr, "    To set bell volume, pitch and duration:\n");
    fprintf (stderr, "\t b [vol [pitch [dur]]]          b on\n");
    fprintf (stderr, "    To turn keyclick off:\n");
    fprintf (stderr, "\t-c                c off               c 0\n");
    fprintf (stderr, "    To set keyclick volume:\n");
    fprintf (stderr, "\t c [0-100]        c on\n");
    fprintf (stderr, "    To set the font path:\n" );
    fprintf (stderr, "\t fp= path[,path...]\n" );
    fprintf (stderr, "    To restore the default font path:\n");
    fprintf (stderr, "\t fp default\n");
    fprintf (stderr, "    To have the server reread font databases:\n");
    fprintf (stderr, "\t fp rehash\n");
    fprintf (stderr, "    To remove elements from font path:\n");
    fprintf (stderr, "\t-fp path[,path...]  fp- path[,path...]\n");
    fprintf (stderr, "    To prepend or append elements to font path:\n");
    fprintf (stderr, "\t+fp path[,path...]  fp+ path[,path...]\n");
    fprintf (stderr, "    To set LED states off or on:\n");
    fprintf (stderr, "\t-led [1-32]         led off\n");
    fprintf (stderr, "\t led [1-32]         led on\n");
    fprintf (stderr, "    To set mouse acceleration and threshold:\n");
    fprintf (stderr, "\t m [acc [thr]]    m default\n");
    fprintf (stderr, "    To set pixel colors:\n");
    fprintf (stderr, "\t p pixel_value color_name\n");
    fprintf (stderr, "    To turn auto-repeat off or on:\n");
    fprintf (stderr, "\t-r     r off        r    r on\n");
    fprintf (stderr, "    For screen-saver control:\n");
    fprintf (stderr, "\t s [timeout [cycle]]  s default\n");
    fprintf (stderr, "\t s blank              s noblank\n");
    fprintf (stderr, "\t s expose             s noexpose\n");
    fprintf (stderr, "    For status information:  q   or  query\n");
    exit(0);
}

error( message )
    char *message;
{
    fprintf( stderr, "%s: %s\n", progName, message );
    exit( 1 );
}


int local_xerror (dpy, rep)
    Display *dpy;
    XErrorEvent *rep;
{
    char dummy[10];
    char *errname = NULL;

    if (rep->request_code == X_StoreColors) {
	switch (rep->error_code) {
	  case BadAccess:
	    errname = "BadAccess, pixel not allocated read/write";
	    break;
	  case BadValue:
	    errname = "BadValue, invalid pixel number";
	    break;
	}
    }

    if (errname) {
	fprintf (stderr, "%s:  error in storing color:  %s\n",
		 progName, errname);
    } else {
	XmuPrintDefaultErrorMessage (dpy, rep, stderr);
    }
    return (0);
}
